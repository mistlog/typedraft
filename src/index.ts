export * from "./core/transcriber";

export * from "./code-object/export-class";
export * from "./code-object/local-context";
export * from "./code-object/inline-context";
export * from "./code-object/method";
export * from "./code-object/module";

export * from "./plug-in/draft-plugin-class";
export * from "./plug-in/draft-plugin-dsl";
export * from "./plug-in/draft-plugin-filter";
export * from "./plug-in/draft-plugin-local-context";
export * from "./plug-in/draft-plugin-refresh";

export * from "./common/utility";

/**
 *
 */
import { Transcriber } from "./core/transcriber";
import { RefreshDraftPlugin } from "./plug-in/draft-plugin-refresh";
import { DSLPlugin } from "./plug-in/draft-plugin-dsl";
import { LocalContextPlugin } from "./plug-in/draft-plugin-local-context";
import { ClassPlugin } from "./plug-in/draft-plugin-class";
import { FilterPlugin } from "./plug-in/draft-plugin-filter";

export function MakeDefaultTranscriber(_module: string) {
    const transcriber = new Transcriber(_module);
    transcriber.m_Plugins = [
        new RefreshDraftPlugin(transcriber),
        new DSLPlugin(transcriber),
        new RefreshDraftPlugin(transcriber),
        new LocalContextPlugin(transcriber),
        new ClassPlugin(transcriber),
        new FilterPlugin(transcriber),
    ];
    return transcriber;
}
