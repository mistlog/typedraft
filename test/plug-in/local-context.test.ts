import { MakeDefaultTranscriber, LocalContextPlugin, RefreshDraftPlugin } from "../../src";

test("get context list: simple", () => {
    //
    const code = `
        <Foo/> + function Test(this: Foo, a: number, b: string){
            <Snippet/>;
            return a.toString()+b;
        };

        function Snippet(){
            console.log("hi");
        }
    `;

    const transcriber = MakeDefaultTranscriber(code);
    transcriber.m_Plugins = [new RefreshDraftPlugin(transcriber)];
    transcriber.Transcribe();

    const plugin = new LocalContextPlugin(transcriber);
    const context_list = plugin.GetContextList();
    expect(context_list).toEqual(["Snippet"]);
});

test("get context list: nested", () => {
    //
    const code = `
        <Foo/> + function Test(this: Foo, a: number, b: string){
            <Snippet/>;
            <AnotherSnippet/>;
            return a.toString()+b;
        };

        function Snippet(){
            console.log("hi");
            <SnippetNested/>;
        }

        function AnotherSnippet(){
            console.log("another");
            <SnippetNested/>;
        }

        function SnippetNested(){
            console.log("nested");
            <DeepSnippet/>;
        }

        function DeepSnippet(){
            console.log("deep");
        }
    `;

    const transcriber = MakeDefaultTranscriber(code);
    transcriber.m_Plugins = [new RefreshDraftPlugin(transcriber)];
    transcriber.Transcribe();

    const plugin = new LocalContextPlugin(transcriber);
    const context_list = plugin.GetContextList();
    expect(context_list).toEqual(["DeepSnippet", "SnippetNested", "Snippet", "AnotherSnippet"]);
});

test("transcribe: simple", () => {
    //
    const code = `
        <Foo/> + function Test(this: Foo, a: number, b: string){
            <Snippet/>;
            return a.toString()+b;
        };

        function Snippet(){
            console.log("hi");
        }
    `;

    const transcriber = MakeDefaultTranscriber(code);
    const plugin = new LocalContextPlugin(transcriber);
    Reflect.set(plugin, "GetContextList", () => ["Snippet"]);
    transcriber.m_Plugins = [new RefreshDraftPlugin(transcriber), plugin];

    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});

test("transcribe: nested", () => {
    //
    const code = `
        <Foo/> + function Test(this: Foo, a: number, b: string){
            <Snippet/>;
            return a.toString()+b;
        };

        function Snippet(){
            console.log("hi");
            <SnippetNested/>;
        }

        function SnippetNested(){
            console.log("nested");
        }
    `;

    const transcriber = MakeDefaultTranscriber(code);
    const plugin = new LocalContextPlugin(transcriber);
    Reflect.set(plugin, "GetContextList", () => ["SnippetNested", "Snippet"]);
    transcriber.m_Plugins = [new RefreshDraftPlugin(transcriber), plugin];

    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});

test("transcribe: export function", () => {
    //
    const code = `
        export function Test(a: number, b: string){
            <Snippet/>;
            return a.toString()+b;
        };

        function Snippet(){
            console.log("export function");
        }
    `;

    const transcriber = MakeDefaultTranscriber(code);
    transcriber.m_Plugins = [
        new RefreshDraftPlugin(transcriber),
        new LocalContextPlugin(transcriber),
    ];

    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});

test("transcribe: method", () => {
    //
    const code = `
        <Foo/> + function Test(this: Foo, a: number, b: string){
            <Snippet/>;
            <AnotherSnippet/>;
            return a.toString()+b;
        };

        function Snippet(){
            console.log("hi");
            <SnippetNested/>;
        }

        function AnotherSnippet(){
            console.log("another");
            <SnippetNested/>;
        }

        function SnippetNested(){
            console.log("nested");
            <DeepSnippet/>;
        }

        function DeepSnippet(){
            console.log("deep");
        }
    `;

    const transcriber = MakeDefaultTranscriber(code);
    transcriber.m_Plugins = [
        new RefreshDraftPlugin(transcriber),
        new LocalContextPlugin(transcriber),
    ];

    const result = transcriber.Transcribe();
    expect(result).toMatchSnapshot();
});
