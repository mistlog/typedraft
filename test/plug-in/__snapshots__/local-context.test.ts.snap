// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transcribe: export function 1`] = `
"export function Test(a: number, b: string) {
  console.log(\\"export function\\");
  return a.toString() + b;
}

function Snippet() {
  console.log(\\"export function\\");
}"
`;

exports[`transcribe: method 1`] = `
"<Foo /> + function Test(this: Foo, a: number, b: string) {
  console.log(\\"hi\\");
  console.log(\\"nested\\");
  console.log(\\"deep\\");
  console.log(\\"another\\");
  console.log(\\"nested\\");
  console.log(\\"deep\\");
  return a.toString() + b;
};

function Snippet() {
  console.log(\\"hi\\");
  console.log(\\"nested\\");
  console.log(\\"deep\\");
}

function AnotherSnippet() {
  console.log(\\"another\\");
  console.log(\\"nested\\");
  console.log(\\"deep\\");
}

function SnippetNested() {
  console.log(\\"nested\\");
  console.log(\\"deep\\");
}

function DeepSnippet() {
  console.log(\\"deep\\");
}"
`;

exports[`transcribe: nested 1`] = `
"<Foo /> + function Test(this: Foo, a: number, b: string) {
  console.log(\\"hi\\");
  console.log(\\"nested\\");
  return a.toString() + b;
};

function Snippet() {
  console.log(\\"hi\\");
  console.log(\\"nested\\");
}

function SnippetNested() {
  console.log(\\"nested\\");
}"
`;

exports[`transcribe: simple 1`] = `
"<Foo /> + function Test(this: Foo, a: number, b: string) {
  console.log(\\"hi\\");
  return a.toString() + b;
};

function Snippet() {
  console.log(\\"hi\\");
}"
`;
